### PR TITLE
Create error reproduction | test(torchlib)

### LINKED ISSUE #564: Create an option to auto create reproduction scripts for failed ops
For failed ops that needed to be debugged in onnx or ort, it would be nice if we can add an option in the test script to autogen a repro script for each failures.

### PR DESCRIPTION
**Pull request by justinchuby**
A tool to create reproduction instructions automatically from tests. To use, set `CREATE_REPRODUCTION_REPORT=1`. E.g.

```
CREATE_REPRODUCTION_REPORT=1 python -m pytest onnxscript/tests/function_libs/torch_lib/ops_test.py -k test_output_match_opinfo__concat_cpu_float32

```

- It will create md files with repro script under the `error_reports` folder. 
- Fixed `atleast_*` function implementations.
- Error reports will be uploaded to artifacts in CI.
- Updated test schema matching logic to consider `tensor(*)` compatible with `seq(tensor(*))`. New tests errors discovered by unskipped tests tracked in: https://github.com/microsoft/onnxscript/issues/960.

Close #564

### CODE REVIEW

**Comment by BowenBao** [comment]
Could you share an example of the md file and repro script in description?

File: .github/workflows/main.yaml [hunk 1]
```diff
@@ -62,7 +62,8 @@ jobs:
       - name: Run tests
         run: nox -t ${{ matrix.nox-tag }} --forcecolor -- -v --cov=onnxscript --cov-report=xml --cov-append --cov-branch -n=auto --junit-xml pytest.xml
         env:
-          CATCH_ORT_SEGFAULT: "${{ matrix.os == 'ubuntu-latest' && '1' || '0' }}"
+          CATCH_ORT_SEGFAULT: "${{ matrix.os != 'windows-latest' && '1' || '0' }}"
+          CREATE_REPRODUCTION_REPORT: "${{ matrix.os != 'windows-latest' && '1' || '0' }}"
```
Comment (BowenBao) [inline .github/workflows/main.yaml]: why disable for windows?

File: onnxscript/tests/function_libs/torch_lib/error_reproduction.py [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+        error_text=error_text,
+        test_name=test_name,
+        short_test_name=short_test_name,
+        reproduction_code=reproduction_code,
+        error_stack=error_stack,
+        sys_info=sys_info,
+    )
+
+    # Turn test name into a valid file name
+    markdown_file_name = f'{test_name.split(".")[-1].replace("/", "-").replace(":", "-")}-{str(time.time()).replace(".", "_")}.md'
```
Comment (BowenBao) [inline onnxscript/tests/function_libs/torch_lib/error_reproduction.py:116]:
```suggestion
    markdown_file_name = f'{short_test_name.replace("/", "-").replace(":", "-")}-{str(time.time()).replace(".", "_")}.md'
```

File: onnxscript/tests/function_libs/torch_lib/error_reproduction.py [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+
+# Uncomment this line to save the model to a file for examination
+# onnx.save_model(onnx_model, "{short_test_name}.onnx")
+
+onnx.checker.check_model(onnx_model)
+session = ort.InferenceSession(onnx_model.SerializeToString(), session_options, providers=("CPUExecutionProvider",))
+
+# Run the model
+for _ in range(N):
+    ort_outputs = session.run(None, ort_inputs)
```
Comment (BowenBao) [inline onnxscript/tests/function_libs/torch_lib/error_reproduction.py:45]: Any plan to extend this script to cover mismatch repro as well?

File: onnxscript/tests/function_libs/torch_lib/error_reproduction.py [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+import numpy as np
+from numpy import array, float16, float32, float64, int32, int64
+import onnx
+import onnxruntime as ort
+
+# Run n times
+N = 1
+
+onnx_model_text = """
+{onnx_model_text}
```
Comment (BowenBao) [inline onnxscript/tests/function_libs/torch_lib/error_reproduction.py:26]: I can see this is more convenient that we have a single repro file. I'm slightly worried when there are large initializers in the model. Perhaps it is fine since this is for op level unittests. Maybe add your considerations on this as NOTE comment.

File: onnxscript/tests/function_libs/torch_lib/error_reproduction.py (hunk 3 above)
Comment (justinchuby) [reply onnxscript/tests/function_libs/torch_lib/error_reproduction.py:45]: Yes but I think I will do it in a separate pr (i.e. for the issue reshape with 0 dim for example)

File: onnxscript/tests/function_libs/torch_lib/error_reproduction.py (hunk 4 above)
Comment (justinchuby) [reply onnxscript/tests/function_libs/torch_lib/error_reproduction.py:26]: Will do.

File: .github/workflows/main.yaml (hunk 1 above)
Comment (justinchuby) [reply .github/workflows/main.yaml]: windows is slow at spawning processes. Turns out macos is not fast either. So just ubuntu for now.

Comment (BowenBao) [reply .github/workflows/main.yaml]: ah I remember this slow process spawning issue lol.

File: onnxscript/tests/function_libs/torch_lib/error_reproduction.py [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+import time
+import traceback
+from typing import Any, Mapping
+
+import numpy as np
+import onnx
+import onnxruntime as ort
+import torch
+
+_REPRODUCTION_TEMPLATE = '''\
```
Comment (BowenBao) [inline onnxscript/tests/function_libs/torch_lib/error_reproduction.py:15]: Any consideration using jinja or libcst for codegen?

**Review by BowenBao — APPROVED** [review]

File: onnxscript/tests/function_libs/torch_lib/error_reproduction.py (hunk 5 above)
Comment (justinchuby) [reply onnxscript/tests/function_libs/torch_lib/error_reproduction.py:15]: I didn’t want to take jinja as an extra dependency. As for libcst I think it’s overkill and obscures the actual generated code at the moment

File: onnxscript/tests/function_libs/torch_lib/error_reproduction.py (hunk 2 above)
Comment (justinchuby) [reply onnxscript/tests/function_libs/torch_lib/error_reproduction.py:116]: Done

File: onnxscript/tests/function_libs/torch_lib/error_reproduction.py (hunk 4 above)
Comment (justinchuby) [reply onnxscript/tests/function_libs/torch_lib/error_reproduction.py:26]: Done
